### PR TITLE
Adding command to force launch context to be a string

### PR DESCRIPTION
**Pull request by hncj811-va**
Story # [API-12091](https://vajira.max.gov/browse/API-12091)

### CODE REVIEW

File: src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js [hunk 1]
```diff
@@ -159,7 +159,7 @@ class TokenHandlerClient {
           );
           for (let key in decodedLaunch) {
             if (!responseBody[key]) {
-              responseBody[key] = decodedLaunch[key];
+              responseBody[key] = String(decodedLaunch[key]);
```
Comment (dbrown-va) [inline src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js:162]: Modifying this section (and actually line 154 above which also would require similar casting) only resolves the /token response payload, however subsequent fetches during token validation would still result in possible integer launch context.

Instead, we want to shift this validation left prior to the saving of the launch context in the first place.

https://github.com/department-of-veterans-affairs/lighthouse-oauth-proxy/blob/master/src/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies/saveDocumentLaunchStrategy.js

This does add a bit of complexity since right now we aren't doing any validation on the provided launch context.

We should check that it is either 

If Base64 encoded json
- If there a key of `patient`, and in that case validate to a string is provided

If Non-base64 encoded json
- Assumed to be the patient, validate that the provided value is a string


If either scenario finds an integer, it should return a 400 error like below.

```
{
    "error": "invalid_launch",
    "error_description": "The provided patient launch must be a string."
}
```

Comment (dbrown-va) [inline src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js:162]: Modifying this section (and actually line 154 above which also would require similar casting) only resolves the /token response payload, however subsequent fetches during token validation would still result in possible integer launch context.

Instead, we want to shift this validation left prior to the saving of the launch context in the first place.

https://github.com/department-of-veterans-affairs/lighthouse-oauth-proxy/blob/master/src/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies/saveDocumentLaunchStrategy.js

This does add a bit of complexity since right now we aren't doing any validation on the provided launch context.

We should check that it is either 

If Base64 encoded json
- If there a key of `patient`, and in that case validate to a string is provided

If Non-base64 encoded json
- Assumed to be the patient, validate that the provided value is a string


If either scenario finds an integer, it should return a 400 error like below.

```
{
    "error": "invalid_launch",
    "error_description": "The provided patient launch must be a string."
}
```

**Review by dbrown-va — DISMISSED** [review]
Need to update to validate the launch context prior to saving for CCG tokens.

File: src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js (hunk 1 above)
Comment (crolarlibertyva) [reply src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js:162]: - oops I may have advised him to the wrong spot.

Comment (hncj811-va) [reply src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js:162]: So instead of the String cast we should throw a 400? 

Also launch should always be an encoding or a JSON like below:

{
  patient: 333
}

Comment (hncj811-va) [reply src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js:162]: In the JSON above throw the error?

Comment (crolarlibertyva) [reply src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js:162]: Yeah, something like

```
throw {
      status: 400,
      error: "invalid_request",
      error_description: "The provided patient launch must be a string",
    };
```

or

```
throw {
      status: 400,
      error: "invalid_launch",
      error_description: "The provided patient launch must be a string",
    };
```

Comment (hncj811-va) [reply src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js:162]: Thanks Larry!

Comment (dbrown-va) [reply src/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClient.js]: Turns out this did not shift the validation far enough left.

An invalid `launch` technically would first have a CCG access token minted, followed by the saveDocumentStrategy call hitting the validation and returning the error.

We need to check if in the TokenHandlerClient itself prior to the token minting is possible without too much churn around the other flows.

File: src/oauthHandlers/authorizeHandler.js [hunk 2]
```diff
@@ -99,6 +103,14 @@ const authorizeHandler = async (
       req.query.scope.split(" ").includes("launch") &&
       req.query.launch
     ) {
+      if (!launchValidation(req.query.launch)) {
+        return {
+          status: 400,
+          error: "invalid_request",
+          error_description:
+            "The provided patient launch must be a strin or base64 encoded json",
```
Comment (dbrown-va) [inline src/oauthHandlers/authorizeHandler.js:111]: Let's go ahead only advertise the going forward option to match the authorization docs.

"The provided patient launch must be a Base64-encoded JSON object." 
https://developer.va.gov/explore/authorization/docs/authorization-code?api=clinical_health

File: src/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies/saveDocumentLaunchStrategy.js [hunk 3]
```diff
@@ -8,25 +8,24 @@ class SaveDocumentLaunchStrategy {
     this.config = config;
   }
   async saveDocumentToDynamo(document, tokens) {
```
Comment (dbrown-va) [inline src/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies/saveDocumentLaunchStrategy.js:10]: Since we moved the validation left of `saveDocumentToDynamo` we can revert these changes back to master.

File: src/utils.js [hunk 4]
```diff
@@ -296,6 +296,30 @@ const rewriteRedirect = (config, request, redirect_uri) => {
   return redirect_uri;
 };
 
+const launchValidation = (launch) => {
+  if (launch === null || launch === "") return false;
+  var base64regex = /^([0-9a-zA-Z+/]{4})*(([0-9a-zA-Z+/]{2}==)|([0-9a-zA-Z+/]{3}=))?$/;
```
Comment (dbrown-va) [inline src/utils.js:301]: Will want this to be a `const` similar to the other regex used in across the utils methods.

File: tests/TokenHandlerTests/saveDocumentLaunchStrategy.test.js [hunk 5]
```diff
@@ -30,24 +30,26 @@ describe("saveDocumentToDynamo tests", () => {
     jest.useRealTimers();
   });
 
-  it("Empty Tokens", async () => {
```
Comment (crolarlibertyva) [inline tests/TokenHandlerTests/saveDocumentLaunchStrategy.test.js:33]: This test should be reverted.

**Review by crolarlibertyva — DISMISSED** [review]

File: tests/TokenHandlerTests/saveDocumentLaunchStrategy.test.js (hunk 5 above)
Comment (hncj811-va) [reply tests/TokenHandlerTests/saveDocumentLaunchStrategy.test.js]: please check out now

**Review by crolarlibertyva — APPROVED** [review]

**Review by crolarlibertyva — APPROVED** [review]
